File: Problem3_EmployeeImportance.java
/***************** Time Complexity: O(n) ***************** */
/***************** Space Complexity: O(n) ***************** */
//Depth First Search
/*
// Employee info
class Employee {
    // It's the unique id of each node;
    // unique id of this employee
    public int id;
    // the importance value of this employee
    public int importance;
    // the id of direct subordinates
    public List<Integer> subordinates;
};
*/
class Solution {
    
    public int getImportance(List<Employee> employees, int id) {
        //base case
        if(employees == null || employees.size() == 0)
            return 0;
        
        //create hashmap for id and employee class 
        HashMap<Integer, Employee> map = new HashMap<>();
        for(Employee e : employees)
            map.put(e.id, e);
        
        return getTotalImp(map, id);
    }
    
    private int getTotalImp(HashMap<Integer, Employee> map, Integer id){
        //get if of the current employee
        Employee current_emp = map.get(id);
        //get the current importance value of the employee
        int total_imp = current_emp.importance;
        //for all subordinates present for the employee
        for(Integer subIds : current_emp.subordinates){
            //add all subordinates importance value
            total_imp += getTotalImp(map, subIds);
            
        }
        
        return total_imp;
    }
}

/************************************************************* */
/***************** Time Complexity: O(n) ***************** */
/***************** Space Complexity: O(n) ***************** */
// Breadth First Search
/*
// Employee info
class Employee {
    // It's the unique id of each node;
    // unique id of this employee
    public int id;
    // the importance value of this employee
    public int importance;
    // the id of direct subordinates
    public List<Integer> subordinates;
};
*/

class Solution {
    public int getImportance(List<Employee> employees, int id) {
        if(employees == null || employees.size()==0)
            return 0;
        //Creating HashMap of id and Emp object
        HashMap<Integer, Employee> map = new HashMap<>();
        //for each employee, create an entry
        for(Employee e : employees){
            map.put(e.id, e);
        }
        //queue is created for storing the subordinates of the employee
        Queue<Integer> q = new LinkedList<>();
        //add the first id in the queue
        q.add(id);
        int result = 0;
        while(!q.isEmpty()){
            //poll id and get its importance to add with result
            int eid = q.poll();
            Employee e = map.get(eid);
            result = result + e.importance;
            //get subordinates and add its importance
            for(int sId : e.subordinates){
                q.add(sId);
            }
        }
        
        return result;
        
    }
}

File: Problem-3_employee_importance.py
# APPROACH  1: BFS
# Time Complexity : O(n), n: number of employee records in given list
# Space Complexity : O(n * max size of employee record), due to hashmap
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None
#
#
# Your code here along with comments explaining your approach
# 1. Store all the employee records as value and id as key in hashmap, st access is O(1)
# 2. Start with record of given id. Add it's impotance to total and add it's children to the queue. 
# 3. Keep doing until, the queue is empty. Return total

"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        
        if employees is None:
            return 0
        
        hashmap_emp, queue, total = {}, deque(), 0
        for emp in employees:
            hashmap_emp[emp.id] = emp
            
        queue.append(id)
        while len(queue) > 0:
            emp_id = queue.popleft()
            total += hashmap_emp[emp_id].importance
            
            for sub_emp_id in hashmap_emp[emp_id].subordinates:
                queue.append(sub_emp_id)
                
        return total
        
        
        
        
# APPROACH 2: DFS
# Time Complexity : O(n), n: number of employee records in given list
# Space Complexity : O(n * max size of employee record), due to hashmap
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None
#
#
# Your code here along with comments explaining your approach
# 1. maintain total as global variable. Store all the employee records as value and id as key in hashmap, st access is O(1)
# 2. Start with record of given id. Run dfs from here. Keep calling its children in depth wise manner recursively. (like only its' first child's all children gets processed 
#    then go to it's second child and so on
# 3. return total

"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    def __init__(self):
        self.total = 0
        self.hashmap_emp = {}
        
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        
        if employees is None:
            return 0
        
        for emp in employees:
            self.hashmap_emp[emp.id] = emp
        
        self.dfs(id)
        return self.total
    
        
    def dfs(self, emp_id):
        self.total += self.hashmap_emp[emp_id].importance
        for sub_emp_id in self.hashmap_emp[emp_id].subordinates:
            self.dfs(sub_emp_id)
        
